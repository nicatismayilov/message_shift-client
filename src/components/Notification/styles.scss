.notification {
  display: flex;
  align-items: center;
  width: 300px;

  border-radius: 10px;
  color: #fff;
  font-size: 16px;
  line-height: 1.5;
  font-family: "Quicksand";
  font-weight: 600;
  padding: 16px;
  box-shadow: 0 5px 10px #00000033;
  margin-bottom: 20px;

  &--success {
    background-color: #4caf50;
  }

  &--info {
    background-color: #2196f3;
  }

  &--warning {
    background-color: #fb8c00;
  }

  &--error {
    background-color: #ff5252;
  }

  & > .row {
    width: 100%;
    margin: 0;

    & > [class*="col"] {
      padding: 0;
    }
  }

  &__icon {
    height: 24px;
    margin-right: 16px;
  }

  &__text {
    text-align: left;
  }

  &__btn {
    border: none;
    border-radius: 50%;
    background-color: transparent;
    cursor: pointer;
    height: 24px;
    width: 24px;

    display: flex;
    justify-content: center;
    align-items: center;

    &:focus {
      outline: none;
    }

    &:hover {
      box-shadow: 0 2px 5px #00000033;
    }
  }
}
