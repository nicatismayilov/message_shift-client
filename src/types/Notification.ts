export interface NotificationIcons {
	successIcon: string;
	infoIcon: string;
	warningIcon: string;
	errorIcon: string;
}

export enum NotificationTypes {
	SUCCESS = "SUCCESS",
	INFO = "INFO",
	WARNING = "WARNING",
	ERROR = "ERROR",
}

export interface Notification {
	id: string;
	type: NotificationTypes;
	message: string;
}
