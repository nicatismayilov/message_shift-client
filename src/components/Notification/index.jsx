import React, { useEffect, memo } from "react";
import { motion } from "framer-motion";

import eventBus from "eventBus";

import renderIcon from "utils/renderIcon";

import successIcon from "assets/img/success.svg";
import infoIcon from "assets/img/info.svg";
import warningIcon from "assets/img/warning.svg";
import errorIcon from "assets/img/error.svg";
import dismissIcon from "assets/img/dismiss.svg";

import "./styles.scss";

const icons = {
	successIcon,
	infoIcon,
	warningIcon,
	errorIcon,
};

const Notification = (props) => {
	const { message, type, id } = props;

	const handleRemove = () => {
		eventBus.dispatch("remove-notification", id);
	};

	const useMountEffect = () => {
		const id = setTimeout(() => {
			handleRemove();
		}, 2000);

		return () => {
			clearTimeout(id);
		};
	};

	useEffect(useMountEffect, []);

	return (
		<motion.li
			initial={{ opacity: 0, scale: 0.8 }}
			animate={{ opacity: 1, scale: 1 }}
			exit={{ opacity: 0, scale: 0, transition: { duration: 0.2 } }}
			layoutTransition={{ duration: 0.2 }}
			className={`notification notification--${type}`}
		>
			<div className='row justify-between align-center'>
				<div className='col-10 flex align-center'>
					<img className='notification__icon' src={renderIcon(type, icons)} alt={type} />

					<div className='notification__text'>{message}</div>
				</div>

				<div className='col-2 flex justify-end'>
					<button className='notification__btn' onClick={handleRemove}>
						<img src={dismissIcon} alt='x' />
					</button>
				</div>
			</div>
		</motion.li>
	);
};

export default memo(Notification);
