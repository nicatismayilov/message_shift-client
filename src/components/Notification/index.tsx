import React, { useEffect, memo } from "react";
import { motion } from "framer-motion";

import eventBus from "eventBus";

import renderIcon from "utils/renderIcon";

import successIcon from "assets/img/success.svg";
import infoIcon from "assets/img/info.svg";
import warningIcon from "assets/img/warning.svg";
import errorIcon from "assets/img/error.svg";
import dismissIcon from "assets/img/dismiss.svg";

import "./styles.scss";

const icons = {
	successIcon,
	infoIcon,
	warningIcon,
	errorIcon,
};

interface Props {
	id: string;
	type: string;
	message: string;
}

const Notification: React.FC<Props> = (props) => {
	const { message, type, id } = props;

	const handleRemoveNorification = (): void => {
		eventBus.dispatch("remove-notification", id);
	};

	const handleOnMount = () => {
		const id = setTimeout(() => {
			handleRemoveNorification();
		}, 3000);

		return () => {
			clearTimeout(id);
		};
	};

	useEffect(handleOnMount, []);

	return (
		<motion.li
			initial={{ opacity: 1, scale: 0 }}
			animate={{
				opacity: 1,
				scale: 1,
				transition: { ease: [0.18, 0.89, 0.43, 1.19], duration: 0.2 },
			}}
			exit={{ opacity: 0, transition: { duration: 0.2 } }}
			layoutTransition={{ duration: 0.2 }}
			className={`notification notification--${type}`}
		>
			<div className='row justify-between align-center'>
				<div className='col-10 d-flex align-center'>
					<img className='notification__icon' src={renderIcon(type, icons)} alt={type} />

					<div className='notification__text'>{message}</div>
				</div>

				<div className='col-2 d-flex justify-end'>
					<button className='notification__btn' onClick={handleRemoveNorification}>
						<img src={dismissIcon} alt='x' />
					</button>
				</div>
			</div>
		</motion.li>
	);
};

export default memo(Notification);
