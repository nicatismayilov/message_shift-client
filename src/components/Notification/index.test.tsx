import React from "react";
import { render, screen } from "@testing-library/react";

import Notification from "./index";

describe("Notification component", () => {
	const mockId = "111111";
	const mockType = "success";
	const mockMessage = "Test Notification";

	it("should match snapshot", () => {
		const { baseElement } = render(
			<Notification id={mockId} type={mockType} message={mockMessage} />
		);

		expect(baseElement).toMatchSnapshot();
	});

	it("should display message correctly", () => {
		render(<Notification id={mockId} type={mockType} message={mockMessage} />);

		const Message = screen.getByTestId("notification-message");
		expect(Message).toBeInTheDocument();
		expect(Message).toHaveTextContent(mockMessage);
	});
});
