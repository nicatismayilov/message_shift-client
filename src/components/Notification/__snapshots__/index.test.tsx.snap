// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component should match snapshot 1`] = `
<body>
  <div>
    <li
      class="notification notification--success"
      data-testid="notification"
      style="opacity: 1; transform: scale(0) translateZ(0);"
    >
      <div
        class="row justify-between align-center"
      >
        <div
          class="col-10 d-flex align-center"
        >
          <img
            alt="success"
            class="notification__icon"
            src=""
          />
          <div
            class="notification__text"
            data-testid="notification-message"
          >
            Test Notification
          </div>
        </div>
        <div
          class="col-2 d-flex justify-end"
        >
          <button
            class="notification__btn"
            data-testid="notification-close-btn"
          >
            <img
              alt="x"
              src="dismiss.svg"
            />
          </button>
        </div>
      </div>
    </li>
  </div>
</body>
`;
